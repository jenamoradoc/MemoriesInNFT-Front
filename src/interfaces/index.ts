export interface LoginResults {
  id: number;
  avatar_url: string;
  login: string;
  type: string;
}

export interface FetchResponse {
  items: LoginResults[];
}
