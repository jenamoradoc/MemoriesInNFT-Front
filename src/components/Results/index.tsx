import { FC, useEffect, useState } from "react";

import { LoginResults } from "../../interfaces";
import { getDataFn } from "../../utils";
import team from "../../assets/team.svg";
import "./results.css";

interface Props {
  login: string;
  results: LoginResults[];
  getResults: (data: LoginResults[]) => void;
  changeLoading: (state: boolean) => void;
  changeMessage: (message: string) => void;
}

interface SelectOption {
  value: string;
  text: string;
}

const Results: FC<Props> = ({
  login,
  results,
  getResults,
  changeLoading,
  changeMessage,
}) => {
  const [page, setPage] = useState<number>(1);

  console.log(results);

  const options: SelectOption[] = [
    { value: "best match", text: "Best Match" },
    { value: "followers", text: "Followers" },
    { value: "repositories", text: "Repositories" },
    { value: "joined", text: "Joined" },
  ];

  const [selected, setSelected] = useState(options[0].value);

  useEffect(() => {
    if (page > 1 && selected.trim() !== "") {
      getDataFn({
        login,
        page,
        setLoading: changeLoading,
        getResults,
        setMessage: changeMessage,
        sort: selected,
      });
    } else if (page === 1 && login.trim() !== "" && selected.trim() !== "") {
      getDataFn({
        login,
        setLoading: changeLoading,
        getResults,
        setMessage: changeMessage,
        sort: selected,
      });
    }

    // eslint-disable-next-line
  }, [page, selected]);

  const pages: number[] =
    results.length === 9 ? [1, 2, 3, 4, 5, 6, 7, 8, 9] : [1];

  return (
    <div>
      {results.length > 0 ? (
        <>
          <select
            className="results__sort"
            value={selected}
            onChange={({ target }) => setSelected(target.value)}
          >
            <option value="" disabled>
              --Choose an option
            </option>
            {options.map(({ value, text }) => (
              <option key={value} value={value}>
                {text}
              </option>
            ))}
          </select>
          <ul className="results__list">
            {results.map(({ id, login, avatar_url, type }) => (
              <li key={id} className="results__list-item">
                <img
                  src={avatar_url}
                  alt={login}
                  className="results__list-item-image"
                  draggable={false}
                />
                <strong className="results__list-item-login">{login}</strong>
                <span className="results__list-item-type">{type}</span>
              </li>
            ))}
          </ul>
          <div className="buttons__container">
            <button
              onClick={() => setPage((page) => page - 1)}
              disabled={page <= 1}
              style={{ borderRadius: "5px 0 0 5px" }}
            >
              &laquo;
            </button>
            {pages.map((pageNumber) => (
              <button
                key={pageNumber}
                onClick={() => setPage(pageNumber)}
                disabled={page === pageNumber}
                data-testid={pageNumber}
              >
                {pageNumber}
              </button>
            ))}
            <button
              onClick={() => setPage((page) => page + 1)}
              disabled={page === pages.length}
              style={{ borderRadius: "0 5px 5px 0" }}
            >
              &raquo;
            </button>
          </div>
        </>
      ) : (
        <div className="results__message">
          {login.trim().length === 0 ? (
            <>
              <h2>GitHub User Search App</h2>
              <img src={team} alt="Team" draggable={false} />
            </>
          ) : results.length === 0 ? (
            <>
              <h2>Results not found!</h2>
              <img src={team} alt="Team" draggable={false} />
            </>
          ) : null}
        </div>
      )}
    </div>
  );
};

export default Results;
