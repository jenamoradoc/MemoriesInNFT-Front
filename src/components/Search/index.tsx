import { FC, FormEvent, useState, useEffect } from "react";

import Spinner from "../Spinner";

import { LoginResults } from "../../interfaces";
import { getDataFn } from "../../utils";
import "./search.css";

interface Props {
  usernameEntered: (login: string) => void;
  getResults: (data: LoginResults[]) => void;
  changeLoading: (state: boolean) => void;
  loading: boolean;
  results: LoginResults[];
  changeMessage: (message: string) => void;
  message: string;
}

const Search: FC<Props> = ({
  usernameEntered,
  getResults,
  loading,
  changeLoading,
  results,
  changeMessage,
  message,
}) => {
  const [login, setLogin] = useState<string>("");

  useEffect(() => {
    if (results.length > 0) {
      getResults(results);
    }

    // eslint-disable-next-line
  }, [results]);

  const handleSubmit = async (
    event: FormEvent<HTMLFormElement>
  ): Promise<void> => {
    event.preventDefault();

    if (login.trim().length === 0) return;

    usernameEntered(login);
    getDataFn({
      login,
      setLoading: changeLoading,
      getResults,
      setMessage: changeMessage,
    });
  };

  const handleClearFilter = () => {
    if (login.trim().length > 0) {
      setLogin("");
      usernameEntered("");
      getResults([]);
    }
  };

  return (
    <>
      <aside className="search__form-container">
        <form onSubmit={handleSubmit} className="search__form">
          <h1 className="search__form-title">Search github user(s)</h1>
          <input
            type="text"
            value={login}
            autoFocus
            onChange={({ target }) => setLogin(target.value)}
            className="search__form-input"
          />

          <div className="search__form-btns__container">
            <button type="submit" className="search__form-btn">
              Submit
            </button>

            <button className="search__form-btn" onClick={handleClearFilter}>
              Clear Filter
            </button>
          </div>

          {message.trim().length > 0 && (
            <p className="error__message">{message}</p>
          )}
        </form>
      </aside>
      {loading && <Spinner />}
    </>
  );
};

export default Search;
