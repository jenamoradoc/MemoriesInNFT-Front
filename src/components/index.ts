import Results from "./Results";
import Search from "./Search";
import Spinner from "./Spinner";

export { Results, Search, Spinner };
