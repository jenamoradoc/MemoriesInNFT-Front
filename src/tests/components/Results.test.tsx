import { render, screen } from "@testing-library/react";

import { Results } from "../../components";
import { LoginResults } from "../../interfaces";

describe("Testing <Results />", () => {
  const changeLoading = jest.fn();
  const getResults = jest.fn();
  const changeMessage = jest.fn();
  const results: LoginResults[] = [];
  const queryLogin: string = "";

  test("should render the initial screen", () => {
    const { container } = render(
      <Results
        changeLoading={changeLoading}
        changeMessage={changeMessage}
        getResults={getResults}
        login={queryLogin}
        results={results}
      />
    );

    expect(container).toMatchSnapshot();
    expect(container).toBeInTheDocument();
  });

  test("should render a list of elements", () => {
    const resultsList: LoginResults[] = [
      {
        avatar_url: "pedro.jpg",
        login: "pedro",
        id: 1234,
        type: "User",
      },
      {
        avatar_url: "pedro123.jpg",
        login: "pedro123",
        id: 12344,
        type: "User",
      },
    ];

    const { container } = render(
      <Results
        changeLoading={changeLoading}
        changeMessage={changeMessage}
        getResults={getResults}
        login={queryLogin}
        results={resultsList}
      />
    );

    expect(container).toMatchSnapshot();
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);
  });
});
