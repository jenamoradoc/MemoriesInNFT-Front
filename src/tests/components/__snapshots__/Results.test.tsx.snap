// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Results /> should render a list of elements 1`] = `
<div>
  <div>
    <select
      class="results__sort"
    >
      <option
        disabled=""
        value=""
      >
        --Choose an option
      </option>
      <option
        value="best match"
      >
        Best Match
      </option>
      <option
        value="followers"
      >
        Followers
      </option>
      <option
        value="repositories"
      >
        Repositories
      </option>
      <option
        value="joined"
      >
        Joined
      </option>
    </select>
    <ul
      class="results__list"
    >
      <li
        class="results__list-item"
      >
        <img
          alt="pedro"
          class="results__list-item-image"
          draggable="false"
          src="pedro.jpg"
        />
        <strong
          class="results__list-item-login"
        >
          pedro
        </strong>
        <span
          class="results__list-item-type"
        >
          User
        </span>
      </li>
      <li
        class="results__list-item"
      >
        <img
          alt="pedro123"
          class="results__list-item-image"
          draggable="false"
          src="pedro123.jpg"
        />
        <strong
          class="results__list-item-login"
        >
          pedro123
        </strong>
        <span
          class="results__list-item-type"
        >
          User
        </span>
      </li>
    </ul>
    <div
      class="buttons__container"
    >
      <button
        disabled=""
        style="border-radius: 5px 0 0 5px;"
      >
        «
      </button>
      <button
        data-testid="1"
        disabled=""
      >
        1
      </button>
      <button
        disabled=""
        style="border-radius: 0 5px 5px 0;"
      >
        »
      </button>
    </div>
  </div>
</div>
`;

exports[`Testing <Results /> should render the initial screen 1`] = `
<div>
  <div>
    <div
      class="results__message"
    >
      <h2>
        GitHub User Search App
      </h2>
      <img
        alt="Team"
        draggable="false"
        src="team.svg"
      />
    </div>
  </div>
</div>
`;
